src/libsystemd/sd-device/sd-device.c

```diff
@@ -1873,13 +1872,33 @@ _public_ int sd_device_set_sysattr_value(sd_device *device, const char *sysattr,
                 return 0;
         }
 
+        value_len = strlen(_value);
+
+        /* drop trailing newlines */
+        while (value_len > 0 && _value[value_len - 1] == '\n')
+                _value[--value_len] = '\0';
+
+        /* value length is limited to 4k */
+        if (value_len > 4096)
+                return -EINVAL;
+
+        value = strdup(_value);
+        if (!value)
+                return -ENOMEM;
+
         r = sd_device_get_syspath(device, &syspath);
         if (r < 0)
                 return r;
 
         path = strjoina(syspath, "/", sysattr);
-        r = lstat(path, &statbuf);
-        if (r < 0) {
+
+        fd = open(path, O_WRONLY | O_CLOEXEC | O_NOFOLLOW);
+        if (fd < 0) {
+                if (errno == ELOOP)
+                        return -EINVAL;
+                if (errno == EISDIR)
+                        return -EISDIR;
+
                 value = strdup("");
```

Comment:
This clobbers the strdup'ed value from above (line 1885) and thus leaks memory. Maybe move the strdup into an else branch of this `if`? Or even easier, do the strdup right before the one call that actually needs it, i. e. `device_add_sysattr_value()`, and drop this and the above strdup?

I also don't quite understand the logic here -- if the `open()` fails, this leaves the real sysfs unchanged, but it calls `device_add_sysattr_value()` with an empty value. Why should the internal state change here if the real state doesn't?